From bbf8911d2b150db686097246f6102ce51db082c6 Mon Sep 17 00:00:00 2001
From: Qnorsten <qnorsten@gmail.com>
Date: Sat, 30 Jul 2022 12:55:44 +0200
Subject: [PATCH] update_engine: handle payload that contain xbl_lp5 partitions

This is an attempt to mimic the modifications done to the update_engine by OnePlus.
Some devices (OnePlus 8T and OnePlus 9R)  can have either LPDDR4X or LPDDR5 ram.
For these devices OnePlus OTA payloads contain both regular xbl/xbl_config images for LPDDR4X devices and additional xbl_lp5/xbl_config_lp5 images for LPDDR5 devices.

For LPDDR4X devices the lp5 partitions aren't flashed.
For LPDDR5 devices the lp5 partitions are flashed to the regular xbl partitions.

On OOS11the prop   ro.boot.ddr_type prop can be used to determine ddr type  (0 = LPDDR4X; 1= LPDDR5)
On OOS12  /proc/devinfo/ddr_type will contain either "Device version: DDR4" or "Device version: DDR5"

Change-Id: I900fc4b3a6ab8d0ce877f4b7347a187cdbb3b123
---

diff --git a/aosp/dynamic_partition_control_android.cc b/aosp/dynamic_partition_control_android.cc
index 538b57c..daf03ee 100644
--- a/aosp/dynamic_partition_control_android.cc
+++ b/aosp/dynamic_partition_control_android.cc
@@ -99,6 +99,52 @@
   Cleanup();
 }
 
+
+//bool TWFunc::Path_Exists(string Path)
+bool DynamicPartitionControlAndroid::Path_Exists(std::string Path) {
+	struct stat st;
+	return stat(Path.c_str(), &st) == 0;
+}
+
+//int TWFunc::read_file(string fn, string& results) 
+int DynamicPartitionControlAndroid::read_file(std::string fn, std::string& results) {
+	std::ifstream file;
+	file.open(fn.c_str(), std::ios::in);
+
+	if (file.is_open()) {
+		std::string line;
+		while (std::getline(file, line)) {
+			results += line;
+		}
+		file.close();
+		return 0;
+	}
+	
+	LOG(INFO) << "Cannot find file" << fn.c_str() << "\n";
+	return -1;
+}
+// OOS11 used ro.boot.ddr_type prop (0 = LPDDR4X; 1= LPDDR5)
+constexpr char kIsDDR5[] = "ro.boot.ddr_type";
+// OOS12 used /proc/devinfo/ddr_type (Device version: DDR4; Device version: DDR5)
+const std::string kDDRType = "/proc/devinfo/ddr_type";
+// Check if device is LPDDR4X or LPDDR5 (Oneplus8T/Oneplus9R)
+bool DynamicPartitionControlAndroid::IsDDR5() {
+	if (Path_Exists(kDDRType)) {
+		std::string ddr_type;
+		if(read_file(kDDRType, ddr_type) != -1) {
+			std::string ddr5 = "DDR5";
+			std::string::size_type i = ddr_type.find(ddr5);
+			if (i != std::string::npos) {
+				return true;
+			}
+			else {
+				return false;
+			}
+		}
+	}	
+	return GetBoolProperty(kIsDDR5, false);
+}
+
 static FeatureFlag GetFeatureFlag(const char* enable_prop,
                                   const char* retrofit_prop) {
   // Default retrofit to false if retrofit_prop is empty.
@@ -1149,6 +1195,35 @@
   }
   // Try static partitions.
   auto static_path = GetStaticDevicePath(device_dir, partition_name_suffix);
+  
+  // Mimic the functionality of the Oneplus update_engine to handle
+  // cases when the OTA payload contain xbl and xbl_config images for both LPDR4X and LPDDR5
+  // only current known cases are Oneplus 8T and Oneplus 9R
+  // these payloads contain two extra LPDDR5 specifc images xbl_lp5 and xbl_config_lp5
+  if (partition_name == "xbl_lp5" || partition_name == "xbl_config_lp5") {
+	  // if (!DDR_type_determined) {
+		   // LOG(ERROR) << "The payload contain both ddr4 and ddr5 images but the ddr type of the device couldn't be determined with certainty.";
+			// return {};
+      // Set target of xbl_lp5 image to regular xbl partitions if device have LPDDR5
+      if (DynamicPartitionControlAndroid::IsDDR5()) {
+	    LOG(INFO) << "DDR type is: DDR5 ";
+        std::string lp5 ="_lp5";
+        std::string::size_type i = partition_name.find(lp5);
+        std::string target_partition_name = partition_name;
+        if (i != std::string::npos) {
+            target_partition_name.erase(i,lp5.length());
+        }
+        const auto& partition_new_name_suffix =target_partition_name + SlotSuffixForSlotNumber(slot);
+        static_path = device_dir.Append(partition_new_name_suffix).value();
+      }
+      else {
+		  LOG(INFO) << "DDR type is: DDR4 ";
+          // Set target for xbl_lp5 images to /dev/null if device have LPDDR4X ram
+          static_path = base::FilePath("/dev/null").value();
+        }
+  }
+  
+  
   if (!DeviceExists(static_path)) {
     LOG(ERROR) << "Device file " << static_path << " does not exist.";
     return {};
diff --git a/aosp/dynamic_partition_control_android.h b/aosp/dynamic_partition_control_android.h
index df91401..d90fefe 100644
--- a/aosp/dynamic_partition_control_android.h
+++ b/aosp/dynamic_partition_control_android.h
@@ -17,6 +17,7 @@
 #ifndef UPDATE_ENGINE_AOSP_DYNAMIC_PARTITION_CONTROL_ANDROID_H_
 #define UPDATE_ENGINE_AOSP_DYNAMIC_PARTITION_CONTROL_ANDROID_H_
 
+#include <fstream>
 #include <memory>
 #include <set>
 #include <string>
@@ -233,6 +234,14 @@
  private:
   friend class DynamicPartitionControlAndroidTest;
   friend class SnapshotPartitionTestP;
+  // Used to check the device ram type is LPDDRX4 or LPDDR5
+  // by reading the value of ro.boot.ddr_type
+  // will default to false if prop is missing
+  bool IsDDR5();
+  int read_file(std::string fn, std::string& results);
+  bool Path_Exists(std::string Path);
+  
+  // bool DDR_type_determined = false;
 
   std::optional<base::FilePath> GetSuperDevice();
 
diff --git a/payload_consumer/filesystem_verifier_action.cc b/payload_consumer/filesystem_verifier_action.cc
index 8b39f6d..f12d0d9 100644
--- a/payload_consumer/filesystem_verifier_action.cc
+++ b/payload_consumer/filesystem_verifier_action.cc
@@ -21,13 +21,13 @@
 #include <sys/stat.h>
 #include <sys/types.h>
 #include <unistd.h>
-
 #include <algorithm>
 #include <cstdlib>
 #include <memory>
 #include <string>
 #include <utility>
 
+#include <android-base/properties.h>
 #include <base/bind.h>
 #include <base/strings/string_util.h>
 #include <brillo/data_encoding.h>
@@ -42,6 +42,8 @@
 
 using brillo::data_encoding::Base64Encode;
 using std::string;
+using android::base::GetBoolProperty;
+
 
 // On a partition with verity enabled, we expect to see the following format:
 // ===================================================
@@ -76,6 +78,51 @@
 
 namespace chromeos_update_engine {
 
+//bool TWFunc::Path_Exists(string Path)
+bool FilesystemVerifierAction::Path_Exists(std::string Path) {
+	struct stat st;
+	return stat(Path.c_str(), &st) == 0;
+}
+
+//int TWFunc::read_file(string fn, string& results) 
+int FilesystemVerifierAction::read_file(std::string fn, std::string& results) {
+	std::ifstream file;
+	file.open(fn.c_str(), std::ios::in);
+
+	if (file.is_open()) {
+		std::string line;
+		while (std::getline(file, line)) {
+			results += line;
+		}
+		file.close();
+		return 0;
+	}
+	
+	LOG(INFO) << "Cannot find file" << fn.c_str() << "\n";
+	return -1;
+}
+// OOS11 used ro.boot.ddr_type prop (0 = LPDDR4X; 1= LPDDR5)
+constexpr char kIsDDR5[] = "ro.boot.ddr_type";
+// OOS12 used /proc/devinfo/ddr_type (Device version: DDR4; Device version: DDR5)
+const std::string kDDRType = "/proc/devinfo/ddr_type";
+// Check if device is LPDDR4X or LPDDR5 (Oneplus8T/Oneplus9R)
+bool FilesystemVerifierAction::IsDDR5() {
+	if (Path_Exists(kDDRType)) {
+		std::string ddr_type;
+		if(read_file(kDDRType, ddr_type) != -1) {
+			std::string ddr5 = "DDR5";
+			std::string::size_type i = ddr_type.find(ddr5);
+			if (i != std::string::npos) {
+				return true;
+			}
+			else {
+				return false;
+			}
+		}
+	}	
+	return GetBoolProperty(kIsDDR5, false);
+}
+
 namespace {
 const off_t kReadFileBufferSize = 128 * 1024;
 constexpr float kVerityProgressPercent = 0.6;
@@ -99,6 +146,18 @@
     return;
   }
   install_plan_.Dump();
+  
+  // Mimic the functionality of the Oneplus update_engine to handle
+  // cases when the OTA payload contain xbl and xbl_config images for both
+  // LPDR4X and LPDDR5 only current known cases are Oneplus 8T and Oneplus 9R
+  // these payloads contain two extra LPDDR5 specifc images xbl_lp5 and xbl_config_lp5
+  for (size_t partitionIndex = 0; partitionIndex < install_plan_.partitions.size(); partitionIndex++) {
+    if (install_plan_.partitions[partitionIndex].name == "xbl_lp5" || install_plan_.partitions[partitionIndex].name == "xbl_lp5_config") {
+        xbllp5PartitionsExist = true;
+        break;
+    }
+  }
+  
   StartPartitionHashing();
   abort_action_completer.set_should_complete(false);
 }
@@ -329,6 +388,7 @@
 
   LOG(INFO) << "Hashing partition " << partition_index_ << " ("
             << partition.name << ") on device " << part_path;
+			
   auto success = false;
   if (IsVABC(partition)) {
     success = InitializeFdVABC(ShouldWriteVerity());
@@ -344,9 +404,37 @@
       LOG(ERROR) << "Cannot hash partition " << partition_index_ << " ("
                  << partition.name
                  << ") because its device path cannot be determined.";
-      Cleanup(ErrorCode::kFilesystemVerifierError);
+      Cleanup(ErrorCode::kFilesystemVerifierError); 
       return;
     }
+	
+  // Mimic the functionality of the Oneplus update_engine to handle
+  // cases when the OTA payload contain xbl and xbl_config images for both
+  // LPDDR4X and LPDDR5 devices only current known cases are Oneplus 8T and Oneplus 9R
+  // the seperate LPDDR5 images are named xbl_lp5 and xbl_config_lp5
+  if (xbllp5PartitionsExist == true) {
+    // Skip hash check for regular xbl and xbl_config partitions if 
+    // The devices have LPDDR5 ram and LPDDR5 specfic partions are included in the payload
+      if (FilesystemVerifierAction::IsDDR5() &&
+          (partition.name == "xbl_config" || partition.name == "xbl")) {
+            LOG(INFO) << "Skip hash verification " << partition_index_ << " for ("
+                    << partition.name << ") because current system is LPDDR5 and LPDDR5 specific xbl_lp5 images exists in payload";
+          partition_index_++;
+          StartPartitionHashing();
+          return;
+        }
+      //Skip hash checking on LPDDR5 specific images if they are included in the payload and current device have LPDDR4X ram
+      else if (!FilesystemVerifierAction::IsDDR5() &&
+          (partition.name == "xbl_config_lp5" || partition.name == "xbl_lp5"))
+      {
+        LOG(INFO) << "Skip hash verification " << partition_index_ << " for ("
+                  << partition.name << ") because current system is non LPDDR5";
+        partition_index_++;
+        StartPartitionHashing();
+        return;
+      }
+  }
+	  
     success = InitializeFd(part_path);
   }
   if (!success) {
diff --git a/payload_consumer/filesystem_verifier_action.h b/payload_consumer/filesystem_verifier_action.h
index 850abda..1ea2049 100644
--- a/payload_consumer/filesystem_verifier_action.h
+++ b/payload_consumer/filesystem_verifier_action.h
@@ -20,6 +20,7 @@
 #include <sys/stat.h>
 #include <sys/types.h>
 
+#include <fstream>
 #include <memory>
 #include <string>
 #include <utility>
@@ -102,7 +103,18 @@
   // Starts the hashing of the current partition. If there aren't any partitions
   // remaining to be hashed, it finishes the action.
   void StartPartitionHashing();
+  
+  // Used to check the device ram type is LPDDRX4 or LPDDR5
+  // by reading the value of ro.boot.ddr_type
+  // will default to false if prop is missing
+  bool IsDDR5();
+  int read_file(std::string fn, std::string& results);
+  bool Path_Exists(std::string Path);
 
+  // Used to indicated if specific LPDRX images exist in the payload
+  // Used in some Oneplus models (Oneplus 8T and Oneplus 9R) that can have either LPDDRX4 or LPDDR5 ram
+  bool xbllp5PartitionsExist = false;
+  
   const std::string& GetPartitionPath() const;
 
   bool IsVABC(const InstallPlan::Partition& partition) const;
